protocols/relay/src/behaviour/handler.rs

```diff
@@ -396,149 +408,173 @@ impl Handler {
-                self.queued_events
-                    .push_back(ConnectionHandlerEvent::NotifyBehaviour(
-                        Event::ReservationReqReceived {
-                            inbound_reservation_req,
-                            endpoint: self.endpoint.clone(),
-                            renewed: self.active_reservation.is_some(),
-                        },
-                    ));
-            }
-            inbound_hop::Req::Connect(inbound_circuit_req) => {
-                self.queued_events
-                    .push_back(ConnectionHandlerEvent::NotifyBehaviour(
-                        Event::CircuitReqReceived {
-                            inbound_circuit_req,
-                            endpoint: self.endpoint.clone(),
-                        },
-                    ));
-            }
+    fn process_inbound_request(
+        &self,
+        io: Stream,
+    ) -> BoxFuture<'static, Result<Event, inbound_hop::FatalUpgradeError>> {
+        let mut substream = Framed::new(io, quick_protobuf_codec::Codec::new(MAX_MESSAGE_SIZE));
+
+        let reservation_duration = self.config.reservation_duration;
+        let max_circuit_duration = self.config.max_circuit_duration;
+        let max_circuit_bytes = self.config.max_circuit_bytes;
+        let endpoint = self.endpoint.clone();
+        let renewed = self.active_reservation.is_some();
+
+        async move {
+            let proto::HopMessage {
+                type_pb,
+                peer,
+                reservation: _,
+                limit: _,
+                status: _,
+            } = substream
+                .next()
+                .await
+                .ok_or(inbound_hop::FatalUpgradeError::StreamClosed)??;
+
+            let event = match type_pb {
+                Type::RESERVE => {
+                    let req = ReservationReq {
+                        substream,
+                        reservation_duration,
+                        max_circuit_duration,
+                        max_circuit_bytes,
+                    };
+
+                    Event::ReservationReqReceived {
+                        inbound_reservation_req: req,
+                        endpoint,
+                        renewed,
+                    }
+                }
+                Type::CONNECT => {
+                    let dst = PeerId::from_bytes(
+                        &peer.ok_or(inbound_hop::FatalUpgradeError::MissingPeer)?.id,
+                    )
+                    .map_err(|_| inbound_hop::FatalUpgradeError::ParsePeerId)?;
+                    let req = CircuitReq { dst, substream };
+
+                    Event::CircuitReqReceived {
+                        inbound_circuit_req: req,
+                        endpoint,
+                    }
+                }
+                Type::STATUS => return Err(inbound_hop::FatalUpgradeError::UnexpectedTypeStatus),
+            };
+
+            Ok(event)
```

Comment:
What do you think of making this an `async fn` within the `inbound_hop` module? That would remove some of the noise in the this already large `handler.rs` module.